src/app/libp2p_helper/src/libp2p_helper/main.go

```diff
@@ -675,40 +676,52 @@ type incomingMsgUpcall struct {
 }
 
 func handleStreamReads(app *app, stream net.Stream, idx int) {
+	// create buffer stream for non-blocking read
+	r := bufio.NewReader(stream)
+
+	var msgBytes []byte
 	go func() {
 		defer func() {
 			_ = stream.Close()
 		}()
-
-		buf := make([]byte, 4096)
 		for {
-			len, err := stream.Read(buf)
-
-			if len != 0 {
-				app.writeMsg(incomingMsgUpcall{
-					Upcall:    "incomingStreamMsg",
-					Data:      codaEncode(buf[:len]),
+			length, err := readLEB128ToUint64(r)
+			if err == io.EOF {
+				continue
+			} else if err != nil {
+				app.writeMsg(streamLostUpcall{
+					Upcall:    "streamLost",
 					StreamIdx: idx,
+					Reason:    fmt.Sprintf("read failure: %s", err.Error()),
 				})
+				return
+			}
+
+			if length == 0 {
+				continue
 			}
 
-			if err != nil && err != io.EOF {
+			msgBytes = make([]byte, length)
+			n, err := io.ReadFull(r, msgBytes)
+			if err != nil {
 				app.writeMsg(streamLostUpcall{
 					Upcall:    "streamLost",
 					StreamIdx: idx,
-					Reason:    fmt.Sprintf("read failure: %s", err.Error()),
+					Reason:    fmt.Sprintf("read failure: %s, read %d bytes", err.Error(), n),
 				})
-				break
```

Comment:
Before, we used to read 1 message per stream, and then exit the `handleStreamReads` loop. In the new workflow, it continuously reads messages from the stream, and will only exit if there is an unexpected error from reading the stream. Is it not reasonable that we could send a special byte to between messages to represent that we are "done" writing messages, rather than waiting for an error to occur when the other side hangs up the stream?

Reply:
if the remote side is done sending messages over this stream, then they should close the stream. otherwise the stream will be left open with neither side reading or writing to it